### PR TITLE
feat(css/lints): Implement `custom-property-no-missing-var-function` rule

### PR DESCRIPTION
**Pull request by g-plane**
**Description:**

Implemented [`custom-property-no-missing-var-function`](https://stylelint.io/user-guide/rules/list/custom-property-no-missing-var-function) rule.

**BREAKING CHANGE:**

No.

### CODE REVIEW

File: crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ctx: LintRuleContext<()>,
+
+    in_declaration: bool,
+    in_declaration_value: bool,
+    in_var_function: bool,
+}
+
+impl Visit for CustomPropertyNoMissingVarFunction {
+    fn visit_declaration(&mut self, declaration: &Declaration) {
+        self.in_declaration = true;
```
Comment (kdy1) [inline crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs:27]: I think we should preserve old state, as declaration can be nested.

Comment (kdy1) [reply crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs:27]: https://docs.rs/swc_css_ast/latest/swc_css_ast/enum.ComponentValue.html#variant.DeclarationOrAtRule
https://docs.rs/swc_css_ast/latest/swc_css_ast/enum.DeclarationOrAtRule.html
https://docs.rs/swc_css_ast/latest/swc_css_ast/struct.Declaration.html

Comment (kdy1) [reply crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs:27]: `ComponentValue::DeclarationOrAtRule(DeclarationOrAtRule::Declaration(Declaration {}))`

File: crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        declaration.visit_children_with(self);
+        self.in_declaration = false;
+    }
+
+    fn visit_component_value(&mut self, component_value: &ComponentValue) {
+        self.in_declaration_value = self.in_declaration;
```
Comment (kdy1) [inline crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs:33]: Same here

File: crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        component_value.visit_children_with(self);
+        self.in_declaration_value = false;
+    }
+
+    fn visit_function(&mut self, function: &Function) {
+        self.in_var_function = function.name.value.eq_str_ignore_ascii_case("var");
```
Comment (kdy1) [inline crates/swc_css_lints/src/rules/custom_property_no_missing_var_function.rs:39]: Same here, as `Function` contains `ComponentValue`

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

**Comment by kdy1** [comment]
@alexander-akait Can you review CSS parts?

File: crates/swc_css_lints/tests/rules/fail/custom-property-no-missing-var-function/input.css [hunk 4]
```diff
@@ -0,0 +1,7 @@
+a { color: --foo; }
+a { color: env(--foo); }
+a { color: color(--foo 0% 0% 0% 0%); }
+a { --foo: red; color: --foo; }
+@property --foo {} a { color: --foo; }
+:root { --foo: pink; }  a { color: --foo, red; }
+:root { --bar: 0; } a { color: --foo(--bar); }
```
Comment (alexander-akait) [inline crates/swc_css_lints/tests/rules/fail/custom-property-no-missing-var-function/input.css:7]: Can we also add test for `background-color: var(--foo, var(--bar))` (i.e. `var` in `var`) and also in other functions, for example - `width: calc(var(--variable-width) + 20px);`

Comment (g-plane) [reply crates/swc_css_lints/tests/rules/fail/custom-property-no-missing-var-function/input.css:7]: For inside other functions like `width: calc(var(--variable-width) + 20px)`, this is already included.

**Review by alexander-akait — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_css_lints

**Pull request merged**